### PR TITLE
Galactic: for_each_callback_group backport

### PR DESCRIPTION
**Pull request by adityapande-1995**
This PR aims takle https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1296 by backporting https://github.com/ros2/rclcpp/pull/1723 to galactic without breaking ABI. To do this, I added a new non virtual function in node_base class, and added static members to it to keep track of a global map of mutexes. This map stores the mutexes for each node_base instance. However, while using the new ```for_each_callback_group``` method, the pointer of type ```NodeBaseInterface``` has to be typecasted to ```NodeBase```. The global map of mutexes is written as a separate class and access to it is protected by an internal mutex.

# TODO 
* General code cleanup, linting issues
* Modify ```static_executor_entities_collector```
* Modify the corresponding test cases

### CODE REVIEW

File: rclcpp/include/rclcpp/node.hpp [hunk 1]
```diff
@@ -70,6 +70,8 @@
 #include "rclcpp/timer.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+typedef std::function<void (rclcpp::CallbackGroup::SharedPtr)> CallbackGroupFunction;
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node.hpp:73]: We should remove this, and instead rely on the version that lives in node_base.hpp.

File: rclcpp/include/rclcpp/node.hpp [hunk 2]
```diff
@@ -153,6 +155,9 @@ class Node : public std::enable_shared_from_this<Node>
   const std::vector<rclcpp::CallbackGroup::WeakPtr> &
   get_callback_groups() const;
 
+  RCLCPP_PUBLIC
+  void for_each_callback_group(const CallbackGroupFunction & func);
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node.hpp:159]: We may as well make this one as much like the one on the `master` branch as possible.  That means copying over the style and the docblock as well (the only thing that should be different in the end is the exact namespace that `CallbackGroupFunction` is in).

File: rclcpp/include/rclcpp/node.hpp [hunk 3]
```diff
@@ -1310,9 +1315,9 @@ class Node : public std::enable_shared_from_this<Node>
   const std::string sub_namespace_;
   const std::string effective_namespace_;
 };
-
 }  // namespace rclcpp
 
+
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node.hpp:1320]: Both of these whitespace changes are unnecessary here.  Just leave them as-is.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 4]
```diff
@@ -18,13 +18,26 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include <unordered_map>
+#include <mutex>
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:22]: Alphabetize these with the ones above.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 5]
```diff
@@ -18,13 +18,26 @@
...
+#include <iostream>
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:23]: Remove this one, as it isn't needed.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 6]
```diff
@@ -18,13 +18,26 @@
...
 #include "rclcpp/context.hpp"
 #include "rclcpp/macros.hpp"
 #include "rclcpp/node_interfaces/node_base_interface.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+#if 1
+#define DPRINT(x) std::cout << "DPRINT: " << x << std::endl;
+#else
+#define DPRINT(x)
+#endif
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:35]: Remove this debug code.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 7]
```diff
@@ -18,13 +18,26 @@
...
+
+typedef std::function<void (rclcpp::CallbackGroup::SharedPtr)> CallbackGroupFunction;
+
+class map_of_mutexes;
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:39]: These should both be in the namespace.  I would suggest putting them into the `rclcpp::node_interfaces` namespace, but other namespaces under `rclcpp` would work as well.

Also, make sure to `#include <functional>` at the top of this file to get `std::function`.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 8]
```diff
@@ -145,4 +164,22 @@ class NodeBase : public NodeBaseInterface
...
+
+    // Methods need to be protected by internal mutex
+    void create_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+    std::shared_ptr<std::mutex> get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+    void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+
+    // Members
+    std::unordered_map<const rclcpp::node_interfaces::NodeBase* , std::shared_ptr<std::mutex> > data;
+    std::mutex internal_mutex;
+
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:182]: Unnecessary blank space.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 9]
```diff
@@ -145,4 +164,22 @@ class NodeBase : public NodeBaseInterface
...
+    map_of_mutexes();
+    ~map_of_mutexes();
+
+    // Methods need to be protected by internal mutex
+    void create_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+    std::shared_ptr<std::mutex> get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+    void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+
+    // Members
+    std::unordered_map<const rclcpp::node_interfaces::NodeBase* , std::shared_ptr<std::mutex> > data;
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:180]:
```suggestion
    std::unordered_map<const rclcpp::node_interfaces::NodeBase*, std::shared_ptr<std::mutex>> data;
```

File: rclcpp/src/rclcpp/executor.cpp [hunk 10]
```diff
@@ -268,21 +268,26 @@ Executor::add_node(rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_pt
   // If the node already has an executor
   std::atomic_bool & has_executor = node_ptr->get_associated_with_executor_atomic();
   if (has_executor.exchange(true)) {
-    throw std::runtime_error("Node has already been added to an executor.");
+    throw std::runtime_error(
+            std::string("Node '") + node_ptr->get_fully_qualified_name() +
+            "' has already been added to an executor.");
```
Comment (clalancette) [inline rclcpp/src/rclcpp/executor.cpp:273]: This is an unrelated change, so let's not put it into this PR.

File: rclcpp/src/rclcpp/executor.cpp [hunk 11]
```diff
@@ -190,14 +191,13 @@ Executor::add_callback_groups_from_nodes_associated_to_executor()
   for (auto & weak_node : weak_nodes_) {
     auto node = weak_node.lock();
     if (node) {
-      auto group_ptrs = node->get_callback_groups();
-      std::for_each(
-        group_ptrs.begin(), group_ptrs.end(),
-        [this, node](rclcpp::CallbackGroup::WeakPtr group_ptr)
+      auto node_base_type = std::dynamic_pointer_cast<rclcpp::node_interfaces::NodeBase>(node);
```
Comment (clalancette) [inline rclcpp/src/rclcpp/executor.cpp:194]: This is a bit of bikeshedding, but I feel like `node_base` is a better name for this variable.  This is the instantiation of a type, not the type itself.  Same goes a few times below.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 12]
```diff
@@ -46,6 +46,20 @@ NodeBase::NodeBase(
   associated_with_executor_(false),
   notify_guard_condition_is_valid_(false)
 {
+
+  // Initialize map_of_mutex object if this is the first NodeBase instance
+  if (this->count_of_instances == 0){
+    this->map_object_ptr = std::make_unique<map_of_mutexes>();
+    this->count_of_instances++;
+    DPRINT("NB constructor- Made a new mm object !");
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:54]: Remove this debug print.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 13]
```diff
@@ -46,6 +46,20 @@ NodeBase::NodeBase(
...
+  }
+  else{
+    DPRINT("NB constructor- mm object already exists")
+  }
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:58]: Remove this debug code.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 14]
```diff
@@ -46,6 +46,20 @@ NodeBase::NodeBase(
   associated_with_executor_(false),
   notify_guard_condition_is_valid_(false)
 {
+
+  // Initialize map_of_mutex object if this is the first NodeBase instance
+  if (this->count_of_instances == 0){
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:51]:
```suggestion
  if (this->count_of_instances == 0) {
```

Comment (clalancette) [reply rclcpp/src/rclcpp/node_interfaces/node_base.cpp:51]: Also, it's not clear to me that this is actually a count.  That is, it is incremented here, but it is never decremented.  This feels more like it is a boolean flag of "has this been initialized yet?".  There is also the secondary problem of what happens if two nodes are created on two different threads and race here.  It might be better to rely on C++ static initialization for this object, though that has some of its own problems.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 15]
```diff
@@ -310,3 +327,64 @@ NodeBase::resolve_topic_or_service_name(
...
+
+// Initialize static variables
+std::unique_ptr<map_of_mutexes> NodeBase::map_object_ptr;
+int NodeBase::count_of_instances;
+
+
+//  ***************** Implementation of map_of_mutexes ***************
+map_of_mutexes::map_of_mutexes()
+{
+  DPRINT("mm constructor called")
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:340]: Remove this DPRINT and all of the ones below.

**Review by clalancette — CHANGES_REQUESTED** [review]
I left a preliminary round of review.  I would also suggest rebasing all of the individual commits into a single commit, adding a useful commit message, and signing it properly (which will make the DCO bot happy).

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp (hunk 14 above)
Comment (adityapande-1995) [reply rclcpp/src/rclcpp/node_interfaces/node_base.cpp:51]: I'll change that to a bool and add appropriate locking around it.

**Comment by adityapande-1995** [comment]
Changes:
* Squashed commits, signed 
* Removed debug prints, macro
* removed ``typedef`` lines, ``CallbackGroupFunction`` now works the same way as in thr olling commit
* Changed variable name to ``node_base``
* ``map_of_mutexes`` now in ``rclcpp::node_interfaces`` namespace
* ``count_of_instances`` changed to boolean flag, protected by a mutex now
* Added docstring for ``for_each_callback_group`` in ``node.cpp`` and ``lifecycle_node.cpp``
* Removed blank spaces

TODO
* Modify tests
* Make changes to ``static_entities_collector``
* general cleanup, linting fixes, ament, uncrustify checks

**Comment by adityapande-1995** [comment]
TODO :
- CI failures in ```test_node.cpp```
- Change test cases to use ``for_each_callback_group``

**Comment by adityapande-1995** [comment]
TODO: Modify ```test_allocator_memory_strategy.cpp``` test to use ```for_each_callback_group```

File: rclcpp/include/rclcpp/node.hpp [hunk 16]
```diff
@@ -70,6 +70,7 @@
 #include "rclcpp/timer.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node.hpp:73]: do we need this new line?

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 17]
```diff
@@ -15,27 +15,42 @@
 #ifndef RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
 #define RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
 
+#include <atomic>
+#include <functional>
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:19]: do we need to include this header?

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 18]
```diff
@@ -15,27 +15,42 @@
...
 
+class map_of_mutexes;
+
 /// Implementation of the NodeBase part of the Node API.
 class NodeBase : public NodeBaseInterface
 {
 public:
   RCLCPP_SMART_PTR_ALIASES_ONLY(NodeBase)
 
+  static std::unique_ptr<map_of_mutexes> map_object_ptr;
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:47]: why an unique_ptr?
Can't this just be an `static map_of_mutexes map_object;`

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 19]
```diff
@@ -15,27 +15,42 @@
...
+  static bool map_init_flag;
+  static std::mutex map_init_flag_mutex;
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:49]: do we need to lazely initialize the map?
If yes, you can use std::call_once() and std::once_flag instead of doing it manually.

File: rclcpp/src/rclcpp/node.cpp [hunk 20]
```diff
@@ -605,3 +605,10 @@ Node::get_node_options() const
 {
   return this->node_options_;
 }
+
+void Node::for_each_callback_group(
+  const node_interfaces::NodeBaseInterface::CallbackGroupFunction & func)
+{
+  auto node_base = std::dynamic_pointer_cast<rclcpp::node_interfaces::NodeBase>(node_base_);
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/node.cpp:612]: this is pretty unsafe, we're not even checking if the resulting ptr is not null.
We don't even need to add a method to `NodeBaseInterface` though, we can just add a free function that works like a thread safe interface to iterate over `get_callback_groups()`.
Of course, that only works if all callers are using the new interface, but that's also a problem if we add a method instead of a free function.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp (hunk 19 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:49]: I don't think we need to lazely initialize the map, it seems to me that it can be initialized before main as all other globals.

**Review by ivanpauno — COMMENTED** [review]
I have a bunch of questions/comments.

File: rclcpp/include/rclcpp/node.hpp (hunk 16 above)
Comment (adityapande-1995) [reply rclcpp/include/rclcpp/node.hpp:73]: will delete

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp (hunk 17 above)
Comment (adityapande-1995) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:19]: will delete

File: rclcpp/src/rclcpp/node.cpp (hunk 20 above)
Comment (adityapande-1995) [reply rclcpp/src/rclcpp/node.cpp:612]: Will add a check to make sure the pointer is not NULL.

I had originally planned for a free function, but I discussed with @clalancette  and he suggested we make a non-virtual method in ```NodeBase``` and typecast ```NodebaseInterface``` members and then use the ```for_each_callback_group``` to make the code is consistent between rolling and the backport. Effectively the free function would do the same, typecast NodebaseInterface to NodeBase and then implement the for_each routine.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp (hunk 18 above)
Comment (adityapande-1995) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:45]: Will do

File: rclcpp/src/rclcpp/node.cpp (hunk 20 above)
Comment (clalancette) [reply rclcpp/src/rclcpp/node.cpp:612]: > We don't even need to add a method to `NodeBaseInterface` though, we can just add a free function that works like a thread safe interface to iterate over `get_callback_groups()`.

Yeah, as @adityapande-1995 points out, I don't think it matters whether we make it a free function or a non-virtual member function.  Either case preserves ABI; making it a non-virtual member function is slightly nicer in that you get `this` for free (rather than having to pass it in).

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp (hunk 19 above)
Comment (clalancette) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:49]: Yeah, I agree.  I think that will be better, and will also get rid of the need for the extra mutex while lazily initializing.

File: rclcpp/src/rclcpp/node.cpp (hunk 20 above)
Comment (ivanpauno) [reply rclcpp/src/rclcpp/node.cpp:612]: yes, the only difference is that the free function alternative doesn't need dynamic casting, as it can receive a reference to the base class directly.

I don't think we're extending `NodeBaseInterface` in another place, so I don't mind much.

Comment (ivanpauno) [reply rclcpp/src/rclcpp/node.cpp:612]: example:

```cpp
void
for_each_calback_group_thread_safe(NodeBaseInterface & base_interface, FunctionCgT fn) {
  auto & mutex = get_cg_mutex_for_base_interface(&base_interface);
  std::lock_guard guard{mutex};
  for (const auto & cg : base_interface->get_callback_groups()) {
    fn(cg);
  }
}
```

In that case no dynamic cast is needed because [get_callback_groups](https://github.com/ros2/rclcpp/blob/e6733bf2c6cd40c8d8d26ec451042247577f0901/rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp#L131) is already a virtual method.

In that case, we will need the `NodeBaseInterface` constructor to register the mutex though (I think now `NodeBase` is doing that).

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp (hunk 18 above)
Comment (adityapande-1995) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:45]: I think that was because ```this->map_object = map_of_mutexes()``` was not allowed as the class member ```std::mutex internal_mutex``` is copyable. To make the map_of_mutex static member ptr to an object, I'll have to change the ```internal_mutex``` as a pointer so that the copy constructor is valid.

Comment (clalancette) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:45]: > I think that was because in `node_base` constructor, `this->map_object = map_of_mutexes()`

But we shouldn't need that at all.  Once we have `map_of_mutexes` in static storage, there is no need to do an assignment to another static within the class; just use the global that already exists.

File: rclcpp/src/rclcpp/node.cpp (hunk 20 above)
Comment (clalancette) [reply rclcpp/src/rclcpp/node.cpp:612]: Ah, I see.  Yeah, that could work more nicely, leveraging the `get_callback_groups` that has to stay here.  @adityapande-1995 this is definitely worth exploring, as getting rid of the `dynamic_cast` will be a lot safer.

Comment (adityapande-1995) [reply rclcpp/src/rclcpp/node.cpp:612]: Will do

Comment (adityapande-1995) [reply rclcpp/src/rclcpp/node.cpp]: Added free function, removed the non virtual one and dynamic pointer casts

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 21]
```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
...
+  void create_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  std::shared_ptr<std::mutex>
+  get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+
+  // Members
+  std::unordered_map<const rclcpp::node_interfaces::NodeBaseInterface *,
+    std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
+};
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:176]: I'll suggest just moving this class up above the `NodeBase` definition.  That way you can get rid of the forward-reference to it, and also just define it in the proper namespace.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 22]
```diff
@@ -310,3 +316,51 @@ NodeBase::resolve_topic_or_service_name(
...
+rclcpp::node_interfaces::map_of_mutexes::~map_of_mutexes()
+{
+}
+
+// For each callback group free function implementation
+void rclcpp::node_interfaces::global_for_each_callback_group(
+  NodeBaseInterface * node_base_interface, const NodeBaseInterface::CallbackGroupFunction & func)
+{
+  auto mutex_ptr = NodeBase::map_object.get_mutex_of_nodebase(node_base_interface);
+  if (mutex_ptr) {
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:356]: When can this condition be false?  I think we need to do "something" if this is false, either printing out an error or throwing an exception (depending on how severe the situation is).

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 23]
```diff
@@ -310,3 +316,51 @@ NodeBase::resolve_topic_or_service_name(
   allocator.deallocate(output_cstr, allocator.state);
   return output;
 }
+
+rclcpp::node_interfaces::map_of_mutexes NodeBase::map_object = map_of_mutexes();
+
+rclcpp::node_interfaces::map_of_mutexes::map_of_mutexes()
+{
+}
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:324]: Since this is the default constructor, we don't need to define this.  Just remove this (and the declaration in the class).

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 24]
```diff
@@ -310,3 +316,51 @@ NodeBase::resolve_topic_or_service_name(
...
+void rclcpp::node_interfaces::map_of_mutexes::delete_mutex_of_nodebase(
+  const rclcpp::node_interfaces::NodeBaseInterface * nodebase)
+{
+  std::lock_guard<std::mutex> guard(this->internal_mutex);
+  this->data.erase(nodebase);
+}
+
+rclcpp::node_interfaces::map_of_mutexes::~map_of_mutexes()
+{
+}
```
Comment (clalancette) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:349]: Same here; just remove the default destructor.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 25]
```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
 }  // namespace node_interfaces
 }  // namespace rclcpp
 
+// Class to hold the global map of mutexes
+class rclcpp::node_interfaces::map_of_mutexes
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:160]: Nit: this can be `final`.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 26]
```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
...
+  // Methods need to be protected by internal mutex
+  void create_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  std::shared_ptr<std::mutex>
+  get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+
+  // Members
+  std::unordered_map<const rclcpp::node_interfaces::NodeBaseInterface *,
+    std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:175]: I think these internal members can be `private`, no?

**Review by clalancette — CHANGES_REQUESTED** [review]
Overall, I think this is getting close.  I've left a few more things to think about.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp (hunk 24 above)
Comment (ivanpauno) [reply rclcpp/src/rclcpp/node_interfaces/node_base.cpp:349]: +1
I would also always stick to the rule of zero or rule of five, and never to the rule of three or any other intermediate solution.
See https://en.cppreference.com/w/cpp/language/rule_of_three#the_rule_of_three_five_zero#rule_of_zero, https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp (hunk 22 above)
Comment (adityapande-1995) [reply rclcpp/src/rclcpp/node_interfaces/node_base.cpp:350]: This is probably not necessary, as the mutex is registered when a new ```NodeBase``` is created and deleted only when the ```NodeBase``` object is destroyed, so this should always return a non-NULL pointer. I've added a ```RCLCPP_ERROR```, though I think this check can be removed.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 27]
```diff
@@ -16,26 +16,52 @@
...
 #include <vector>
 
 #include "rcl/node.h"
+#include "rclcpp/callback_group.hpp"
 #include "rclcpp/context.hpp"
 #include "rclcpp/macros.hpp"
 #include "rclcpp/node_interfaces/node_base_interface.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:31]: Nit: unnecessary whitespace change.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 28]
```diff
@@ -16,26 +16,52 @@
...
+{
+public:
+  // Methods need to be protected by internal mutex
+  void create_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+  std::shared_ptr<std::mutex>
+  get_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+
+private:
+  // Members
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:52]: Nit: no need for this comment, it is pretty obvious.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 29]
```diff
@@ -16,26 +16,52 @@
...
+  std::unordered_map<const NodeBaseInterface *, std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
```
Comment (clalancette) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:54]: Nit: all of our member variables end with `_`, so this should be:

```suggestion
  std::unordered_map<const NodeBaseInterface *, std::shared_ptr<std::mutex>> data_;
  std::mutex internal_mutex_;
```

(obviously you'll have to update the code as well)

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp (hunk 22 above)
Comment (clalancette) [reply rclcpp/src/rclcpp/node_interfaces/node_base.cpp:350]: I'd say let's remove the check, then.

**Review by clalancette — CHANGES_REQUESTED** [review]
I have a bunch of nits inline.  Once those are all fixed, I think this is good to go from my perspective, though I'd like to get @ivanpauno 's final thoughts before we run CI on this.

**Review by clalancette — APPROVED** [review]
This looks good to me with three things:

1. Green CI.
2. Running `abi-compliance-checker` before and after this PR to make sure we aren't changing ABI.
3. An approval from @ivanpauno and @cottsay

**Comment by adityapande-1995** [comment]
CI Jobs:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15059/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9792/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12745/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15226/)

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp [hunk 30]
```diff
@@ -46,6 +48,9 @@ NodeBase::NodeBase(
   associated_with_executor_(false),
   notify_guard_condition_is_valid_(false)
 {
+  // Generate a mutex for this instance of NodeBase
+  this->map_object.create_mutex_of_nodebase(this);
```
Comment (cottsay) [inline rclcpp/src/rclcpp/node_interfaces/node_base.cpp:52]: Is it standard practice to refer to static members using the instance reference like this? It seems misleading to me.

File: rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp [hunk 31]
```diff
@@ -89,6 +89,7 @@
 #include "rclcpp_lifecycle/transition.hpp"
 #include "rclcpp_lifecycle/visibility_control.h"
 
+
```
Comment (cottsay) [inline rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp:92]: Any particular reason for this spurious newline?

**Review by cottsay — APPROVED** [review]
Looks good from an ABI compatibility perspective. I had a couple of other comments inline, but nothing critical.

File: rclcpp/src/rclcpp/node_interfaces/node_base.cpp (hunk 30 above)
Comment (adityapande-1995) [reply rclcpp/src/rclcpp/node_interfaces/node_base.cpp]: Changed to use ```NodeBase::``` instead of ```this->```

File: rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp (hunk 31 above)
Comment (adityapande-1995) [reply rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp]: Removed

**Comment by adityapande-1995** [comment]
CI Jobs:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15063/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9795/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12748/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15229/)

**Comment by adityapande-1995** [comment]
@cottsay How should we handle the unstable windows CI ? The failures are in ```rosbag2_transport``` which doesn't seem to be related to this PR.

**Comment by cottsay** [comment]
> How should we handle the unstable windows CI ? The failures are in rosbag2_transport which don't seem to be related to this PR.

If you're confident that they aren't a regression caused by this PR, then I think it's OK to ignore the test failures. It might be worth triggering a baseline build to be sure, though.

**Comment by adityapande-1995** [comment]
CI on raw galactic:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15072/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9804/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12757/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15239/)

**Review by ivanpauno — APPROVED** [review]
LGTM!

**Comment by ivanpauno** [comment]
> @cottsay How should we handle the unstable windows CI ? The failures are in rosbag2_transport which don't seem to be related to this PR.

I can't tell if those failures are related or not, +1 to compare with a baseline build.

**Comment by adityapande-1995** [comment]
> CI on galactic branch of rclcpp:
> 
> * Linux [Build Status](http://ci.ros2.org/job/ci_linux/15072/)
> * Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9804/)
> * macOS [Build Status](http://ci.ros2.org/job/ci_osx/12757/)
> * Windows [Build Status](http://ci.ros2.org/job/ci_windows/15239/)

So this CI was run by changing the branch of rclcpp to ```galactic```, i.e without my changes and it produced the same error in Windows. @cottsay @ivanpauno

**Comment by ivanpauno** [comment]
> So this CI was run by changing the branch of rclcpp to galactic, i.e without my changes and it produced the same error in Windows. @cottsay @ivanpauno

LGTM!

**Review by cottsay — APPROVED** [review]
Thanks for double-checking!

**Comment by adityapande-1995** [comment]
ABI check report

image
